apis/status/v1beta1/util.go

```diff
@@ -0,0 +1,74 @@
+package v1beta1
+
+import (
+	"strings"
+	"sync"
+)
+
+var (
+	podOwnershipEnabled = true
+	ownerMutex          = sync.RWMutex{}
+)
+
+// DisablePodOwnership disables setting the owner reference for Status resource.
+// This should only be used for testing, where a Pod resource may not be available
+func DisablePodOwnership() {
+	ownerMutex.Lock()
+	defer ownerMutex.Unlock()
+	podOwnershipEnabled = false
+}
+
+func PodOwnershipEnabled() bool {
+	ownerMutex.RLock()
+	defer ownerMutex.RUnlock()
+	return podOwnershipEnabled
+}
+
+// dashExtractor unpacks the status resource name, unescaping `-`
+func dashExtractor(val string) []string {
+	b := strings.Builder{}
+	var tokens []string
+	var prevDash bool
+	for _, chr := range val {
+		if prevDash && chr != '-' {
+			tokens = append(tokens, b.String())
+			b.Reset()
+			prevDash = false
+		}
+		if chr == '-' {
+			if prevDash {
+				b.WriteRune(chr)
+				prevDash = false
+				continue
+			} else {
+				prevDash = true
+				continue
+			}
+		}
+		b.WriteRune(chr)
+	}
+	tokens = append(tokens, b.String())
+	return tokens
+}
+
+// dashPacker puts a list of strings into a dash-separated format. Note that
+// it cannot handle empty strings, as that makes the dash separator for the empty
+// string reduce to an escaped dash. This is fine because none of the packed strings
+// are allowed to be empty. If this changes in the future, we could create a placeholder
+// for the empty string, say `b`, and replace all instances of `b` in the input
+// stream with `bb`, which could then be unfolded. If we need that, we are already
+// changing the schema of the status resource, and therefore don't need to deal with
+// it now. It also doesn't handle the case where a value begins or ends with a dash,
+// which is also disallowed by the schema (and would require an additional placeholder
+// character to fix). Finally, note that it is impossible to distinguish between
+// a nil list of strings and a list of one empty string.
+func dashPacker(vals ...string) string {
```

Comment:
It could, though that would complicate the function signatures of the calling functions unnecessarily. I definitely don't like that golang has no exception handling, as it leads to many more if/else statements that make programs less clear.

Do you think there is significant benefit in having this sanity check and that relying on GVK conformance is insufficient?

Reply:
I don't have scenarios in mind when I'd expect this to fail - my comment was around making such errors discoverable in the unlikely event that they occur in the future.

Reply:
Fair, and better safe than sorry. Added the sanity checks.